compiler/rustc_const_eval/src/check_consts/qualifs.rs

```diff
@@ -59,19 +62,9 @@ pub trait Qualif {
     /// It also determines the `Qualif`s for primitive types.
     fn in_any_value_of_ty<'tcx>(cx: &ConstCx<'_, 'tcx>, ty: Ty<'tcx>) -> bool;
 
-    /// Returns `true` if this `Qualif` is inherent to the given struct or enum.
-    ///
-    /// By default, `Qualif`s propagate into ADTs in a structural way: An ADT only becomes
-    /// qualified if part of it is assigned a value with that `Qualif`. However, some ADTs *always*
-    /// have a certain `Qualif`, regardless of whether their fields have it. For example, a type
-    /// with a custom `Drop` impl is inherently `NeedsDrop`.
-    ///
-    /// Returning `true` for `in_adt_inherently` but `false` for `in_any_value_of_ty` is unsound.
-    fn in_adt_inherently<'tcx>(
-        cx: &ConstCx<'_, 'tcx>,
-        adt: AdtDef<'tcx>,
-        args: GenericArgsRef<'tcx>,
-    ) -> bool;
+    /// Returns `true` if the `Qualif` is not structural, i.e. that we should not recurse
+    /// into the operand.
```

Comment:
But then what do we do instead?

I don't think I understand why you are changing the API here.
For `NeedsNonConstDrop`, the behavior in `in_adt_inherently` that I would expect i:
- check if this type has a `Drop` impl at all. If no, we can propagate structurally, so return `false`.
- check if the type's `Drop` impl is `const`. If yes, we can propagate structurally, so return `false`.
- else return `true`, we found a non-const drop.

Reply:
I think there might be issues with propagating structurally even if there is a const drop implementation, since that drop implementation may have side-effects. I'll think of an example when I properly wake up.

Reply:
"qualif" is always an underapproximation: if it returns `true`, that means "maybe". If it returns `false`, that means "definitely not".

So when hitting a struct type, there are only two options:
- this type definitely does not qualify ("does not have the property we are checking for") on its own, but we need to check the fields. Return `false` from `in_adt_inherently`.
- this type might qualify on its own. Return `true`, which will propagate upwards and finish the computation, answering `true` overall.

Currently there is no way to encode "this type definitely does not qualify, no need to check the fields". But I don't think we have such a property.

Reply:
> there might be issues with propagating structurally even if there is a const drop implementation, since that drop implementation may have side-effects

That's why we must always recurse -- unless we are already sure that this type is `NeedsNonConstDrop`, in which case we can stop.

"Qualifying" is usually a *bad* thing: to qualify for `NeedsNonConstDrop` means that we need non-const drop, and therefore we'll show an error during const-checking.

Reply:
The whole point is that we *cannot* recurse when there is a `const Drop` implementation, even if that implementation is const and the operand's fields are structurally const drop, since because the const drop implementation can have side effects, this relation is no longer structural at all. I'm imagining something like this:

```
#![feature(const_trait_impl)]
#![feature(const_precise_live_drops)]

struct NonConstDrop;

struct Foo(Option<NonConstDrop>);

impl Drop for NonConstDrop { fn drop(&mut self) {
    println!("drop");
} }

impl const Drop for Foo {
    fn drop(&mut self) {
        std::mem::forget(self.0.replace(NonConstDrop));
    }
}

fn test() {
    let _ = Foo(None);
}

fn main() {}
```

Reply:
The answer to your original question of:

> But then what do we do instead?

Is just if we determine that due to the ADT that the qualif is non-structural, we fall back to checking `in_any_value_of` and just return true instead of recursing structurally into the value at all, which is what this PR implements. 

This should make sense: after a const `Drop` impl has executed, we must check that *any* value of the type may have a non-const drop, which is achieved by checking if `~const Destruct` may not hold.

Reply:
> The whole point is that we cannot recurse when there is a const Drop implementation, even if that implementation is const and the operand's fields are structurally const drop, since because the const drop implementation can have side effects, this relation is no longer structural at all. I'm imagining something like this:

I don't follow.

I would say in that example, `Foo` qualifies for "needs non-const drop". Remember that this is a "maybe" answer. Analyzing the destructor seems to be asking a bit too much so this is an acceptable overapproximation I would say.

Reply:
Oh, you mean that the structural thing will give the wrong answer since it is structural on the *value*, not the type... I see.

Reply:
> Is just if we determine that due to the ADT that the qualif is non-structural, we fall back to checking in_any_value_of and just return true instead of recursing structurally into the value at all, which is what this PR implements.

It's not obvious to me that this will be the right behavior for all existing qualifs, so it needs careful checking -- and the doc comment needs to explain this as well.

Reply:
Correct. `is_non_structural` means the qualif is not structural in the *value*. 

This is a no-op for the other two qualifs, since for both of them, if they return true here they'll return true in `in_any_value_of_ty`.

Reply:
We already have `deref_structural` below, so the name and/or docs of this function should some clarify how this is a different kind of "structural".

Reply:
Well, alternatively we should just remove `deref_structural` or rename it. It seems to be saving space for something that isn't happening. But I can add a comment, and if you have a better name idea for things then go ahead.

Reply:
>  It seems to be saving space for something that isn't happening.

I don't know what you mean by space.

`deref_structural` was added because originally qualifs assumed that deref was always structural, and I found that to be unsound. So I extended the system to make this a per-qualif option. It is possible that the only qualif that used this is gone now. I am fine with just hard-coding its value to be `false`.

However, `is_non_structural` is still a confusing name. It should be `is_non_structural_adt` or so.

Reply:
Well if you check the code, I renamed it to `is_structural_in_operand` and flipped the polarity.

I'll remove `deref_structural` since it's always hard-coded to false in all qualifs today.

Reply:
`is_structural_in_adt` would IMO be better. It's important to say that this does *not*, e.g., say whether this is structural for a deref.

"in operand" says very little?